### ISSUE TITLE
Wrong return type for `dict.fromkeys`

### ISSUE THREAD

**Issue by Kaniee** [microsoft/pyright#5413]
**Describe the bug**
When creating a `dict` with a specified type using `dict.fromkeys` the type of the dict values is `Any | None` instead of the specified type.

**To Reproduce**
```
dict_optional: dict[int, int | None] = dict.fromkeys(range(10), None)
reveal_type(dict_optional)
dict_int: dict[int, int] = {}
for key, value in dict_optional.items():
    if value is None:
        continue
    reveal_type(value)
    dict_int[key] = value
```

**Expected behavior**
No typing error.
```
information: Type of "dict_optional" is "dict[int, int | None]"
information: Type of "value" is "int"
```

Similar to the mypy output:
```
note: Revealed type is "builtins.dict[builtins.int, Union[builtins.int, None]]"
```

**Output**
```
information: Type of "dict_optional" is "dict[int, Any | None]"
information: Type of "value" is "Any"
```

**VS Code extension or command-line**
Running pyright from the command-line.

Python 3.10.6
pyright 1.1.311

**Additional context**
Type checking mode: strict

**Issue research**
I found some related (closed) issues in typeshed and mypy repos. But I'm not sure what to take from there:
https://github.com/python/typeshed/issues/5512
https://github.com/python/typeshed/pull/3799
https://github.com/python/mypy/pull/8474

_Edited_
Initially I forgot the if-clause so my output also had this error message:
```
error: Argument of type "Any | None" cannot be assigned to parameter "__value" of type "int" in function "__setitem__"
    Type "Any | None" cannot be assigned to type "int"
      Type "None" cannot be assigned to type "int" (reportGeneralTypeIssues)
```

**Comment by GugelRobin** [comment]
Although it's wrong that `value` has type `Any | None` pyright is correct in giving an error, as `value` is actually `int | None` which is invalid for `dict_int` as that accepts only `int` for `__value` in `__setitem__`.

Or did I misunderstand something on why you wouldn't expect an error here? I would...

**Comment by Kaniee** [comment]
Thanks @GugelRobin I accidentally removed the `if value is None continue` part when creating the MWE. So the error is not present anymore.
I updated my example and the the output accordingly.

**Comment by erictraut** [comment]
This issue is related to the heuristics pyright uses for type narrowing upon assignment when the assigned type includes an `Any`. I agree that your code sample demonstrates an inconsistency in the way pyright was handling this case. I've updated the heuristic so it's more consistent. This may create some churn for some typed code bases that use strict typing.

This change will be in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.319, which I just published. It will also be included in a future release of pylance.
